### PR TITLE
Integration Tests - Nested Commenting Feature

### PR DESCRIPTION
**Pull request by thisaltennakoon**

### CODE REVIEW

**Review by tmkasun — APPROVED** [review]

File: modules/integration/tests-common/clients/store/src/gen/java/org/wso2/am/integration/clients/store/api/v1/dto/CommentDTO.java [hunk 1]
```diff
@@ -46,10 +47,77 @@ public class CommentDTO {
         @SerializedName(SERIALIZED_NAME_CREATED_BY)
             private String createdBy;
 
+        public static final String SERIALIZED_NAME_UPDATED_TIME = "updatedTime";
```
Comment (senthuran16) [inline modules/integration/tests-common/clients/store/src/gen/java/org/wso2/am/integration/clients/store/api/v1/dto/CommentDTO.java:50]: This `CommentDTO` class has some bad formatting issues (not indented properly). Shall we fix those?

File: modules/integration/tests-common/clients/store/src/gen/java/org/wso2/am/integration/clients/store/api/v1/dto/PostRequestBodyDTO.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        
+        this.content = content;
+        return this;
+        }
+
+    /**
+        * Content of the comment 
+    * @return content
+    **/
+      @ApiModelProperty(example = "This is a comment", required = true, value = "Content of the comment ")
```
Comment (senthuran16) [inline modules/integration/tests-common/clients/store/src/gen/java/org/wso2/am/integration/clients/store/api/v1/dto/PostRequestBodyDTO.java:51]: This `PostRequestBodyDTO ` class has some bad formatting issues (not indented properly). Shall we fix those?

File: modules/integration/tests-common/integration-test-utils/src/main/java/org/wso2/am/integration/test/impl/RestAPIPublisherImpl.java [hunk 3]
```diff
@@ -1687,4 +1694,159 @@ public HttpResponse deleteAPIProductRevision(String apiUUID, String revisionUUID
...
+     * Add comment to given API
+     *
+     * @param apiId   - api Id
+     * @param comment - comment to  add
+     * @param category - category of the comment
+     * @param replyTo - comment id of the root comment to add replies
+     * @return - http response of add comment
+     * @throws ApiException - throws if add comment fails
+     */
+
```
Comment (senthuran16) [inline modules/integration/tests-common/integration-test-utils/src/main/java/org/wso2/am/integration/test/impl/RestAPIPublisherImpl.java:1708]: Let's remove this unnecessary newline

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/other/PublisherCommentTest.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
+/*
+ *Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (senthuran16) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/other/PublisherCommentTest.java:2]: Let's change the year to `2021`

Comment (thisaltennakoon) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/other/PublisherCommentTest.java:2]: This has been fixed already

File: modules/integration/tests-common/clients/store/src/gen/java/org/wso2/am/integration/clients/store/api/v1/dto/CommentDTO.java (hunk 1 above)
Comment (thisaltennakoon) [reply modules/integration/tests-common/clients/store/src/gen/java/org/wso2/am/integration/clients/store/api/v1/dto/CommentDTO.java:50]: Fixed in https://github.com/wso2/product-apim/pull/13178

File: modules/integration/tests-common/clients/store/src/gen/java/org/wso2/am/integration/clients/store/api/v1/dto/PostRequestBodyDTO.java (hunk 2 above)
Comment (thisaltennakoon) [reply modules/integration/tests-common/clients/store/src/gen/java/org/wso2/am/integration/clients/store/api/v1/dto/PostRequestBodyDTO.java:51]: Fixed in https://github.com/wso2/product-apim/pull/13178

File: modules/integration/tests-common/integration-test-utils/src/main/java/org/wso2/am/integration/test/impl/RestAPIPublisherImpl.java (hunk 3 above)
Comment (thisaltennakoon) [reply modules/integration/tests-common/integration-test-utils/src/main/java/org/wso2/am/integration/test/impl/RestAPIPublisherImpl.java:1708]: Fixed in https://github.com/wso2/product-apim/pull/13178

File: modules/integration/tests-common/integration-test-utils/src/main/java/org/wso2/am/integration/test/impl/RestAPIPublisherImpl.java [hunk 5]
```diff
@@ -1687,4 +1694,159 @@ public HttpResponse deleteAPIProductRevision(String apiUUID, String revisionUUID
...
+    /**
+     * Add comment to given API
+     *
+     * @param apiId   - api Id
+     * @param comment - comment to  add
+     * @param category - category of the comment
+     * @param replyTo - comment id of the root comment to add replies
+     * @return - http response of add comment
+     * @throws ApiException - throws if add comment fails
+     */
```
[the remark this answers is not in the document]
Comment (thisaltennakoon) [reply modules/integration/tests-common/integration-test-utils/src/main/java/org/wso2/am/integration/test/impl/RestAPIPublisherImpl.java:1707]: Fixed in https://github.com/wso2/product-apim/pull/13178

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/other/PublisherCommentTest.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+            assertEquals(reply.getParentCommentId(), rootCommentIdToAddReplies, "Comments do not match");
+            assertEquals(replies.contains(reply.getId()), true, "Comments do not match");
+        }
+
+        // Get  all the comments of  API
+        HttpResponse getCommentsResponse = restAPIPublisher.getComments(apiId, gatewayContextWrk.getContextTenant()
+                .getDomain(),  false, 5, 0);
+        assertEquals(getCommentsResponse.getResponseCode(), Response.Status.OK.getStatusCode(),
+                "Error retrieving comment");
+        Gson getCommentsGson = new Gson();
```
[the remark this answers is not in the document]
Comment (thisaltennakoon) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/other/PublisherCommentTest.java:170]: Fixed in https://github.com/wso2/product-apim/pull/13178

**Pull request merged**